File: leetcode_cpp/707_设计链表.cpp
// 写不出来！！！！！


//struct ListNode {
//    int val;
//    ListNode* next;
//    explicit ListNode(int x) : val(x), next(nullptr) {};
//};
//
//class MyLinkedList {
//    int size;
//    ListNode* head;
//public:
//    MyLinkedList() {
//        size = 0;
//        head = new ListNode(0);
//    }
//
//    int get(int index) {
//        ListNode* cur = head;
//        int cur_index = 0;
//        cur = cur->next;
//
//        if (index >= size || index < 0) {
//            return -1;
//        }
//
//        while (cur_index != index) {
//            if (cur == nullptr) {
//                return -1;
//            }
//            cur = cur->next;
//            cur_index++;
//        }
//        if (cur)
//            return cur->val;
//        else
//            return -1;
//    }
//
//    void addAtHead(int val) {
//        ListNode* element = new ListNode(val);
//        element->next = head->next;
//        head->next = element;
//        size++;
//    }
//
//    void addAtTail(int val) {
//        ListNode* cur = head;
//        // 遍历到尾节点
//        while (cur->next != nullptr) {
//            cur = cur->next;
//        }
//        ListNode* node = new ListNode(val);
//        cur->next = node;
//        size++;
//    }
//
//    void addAtIndex(int index, int val) {
//        ListNode* cur = head;
//        int cur_index = 0;
//        cur = cur->next;
//
//        if (index == 0) {
//            addAtHead(val);
//            size++;
//            return;
//        }
//        if (index == size) {
//            addAtTail(val);
//            size++;
//            return;
//        }
//        if (index > size || index < 0) {
//            return;
//        }
//
//        while (cur_index != index - 1 && cur != nullptr) {
//            cur = cur->next;
//            cur_index++;
//        }
//        ListNode* node = new ListNode(val);
//        if (cur) {
//            node->next = cur->next;
//            cur->next = node;
//            size++;
//        }
//    }
//
//    void deleteAtIndex(int index) {
//        ListNode* cur = head;
//        int cur_index = 0;
//        while (cur_index != index && cur) {
//            cur = cur->next;
//            cur_index++;
//        }
//        if (cur) {
//            if (cur->next) {
//                cur->next = cur->next->next;
//                size--;
//            }
//        }
//    }
//};

File: leetcode_rust/src/707_设计链表.rs
// 写不出来！！！！！！

pub struct MyLinkedList {
    val: i32,
    next: Option<Box<MyLinkedList>>
}
/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyLinkedList {
    fn new() -> Self {
        MyLinkedList {val: 0, next: None}
    }

    fn get(&self, index: i32) -> i32 {
        let mut cur_index = 0;
        let mut cur = &self.next;
        while let Some(node) = cur{
            if cur_index != index {
                cur = &node.next;
            }
        }
        if let Some(node) = cur{
            node.val
        } else {
            -1
        }
    }

    fn add_at_head(&mut self, val: i32) {
        let new_node = Box::new(MyLinkedList{val: val, next: self.next.take()});
        self.next = Some(new_node);
    }

    fn add_at_tail(&mut self, val: i32) {
        let mut cur = &mut self.next;
        let new_node = Box::new(MyLinkedList {val: val, next: None});
        // 一直遍历，直到最后节点位置
        while let Some(node) = cur{
            cur = &mut node.next;
        }
        *cur = Some(new_node);
    }

    fn add_at_index(&mut self, index: i32, val: i32) {
        if index == 0 {
            self.add_at_head(val);
        } else {
            let mut cur_index = 0;
            let mut cur = &mut self.next;


            // 一直遍历到index的前一个节点
            while let Some(nxt) = cur{
                if cur_index != index - 1 {
                    let mut new_node = Box::new(MyLinkedList { val: val, next: None});
                    new_node.next = nxt.next.take();
                    nxt.next = Some(new_node);
                }
                cur = &mut nxt.next;
                cur_index += 1;
            }
        }
    }

    fn delete_at_index(&mut self, index: i32) {
        if index < 0 {
            return;
        }

        let mut i = 0;
        let mut cur = self;
        while let Some(node) = cur.next.take() {
            if i == index {
                cur.next = node.next;
                break;
            }
            i += 1;
            cur.next = Some(node);
            cur = cur.next.as_mut().unwrap();
        }
    }
}